File: java/patterns/src/main/java/org/prototypepatterns/structural/proxy/ProxyPattern.java
package org.prototypepatterns.structural.proxy;

public final class ProxyPattern {
    private ProxyPattern() {
    }

    public static final class RealImage {
        private static int loadCount;
        private final String path;

        public RealImage(String path) {
            this.path = path;
            loadCount++;
        }

        public String display() {
            return "display:" + path;
        }

        public static int loadCount() {
            return loadCount;
        }

        public static void reset() {
            loadCount = 0;
        }
    }

    public static final class ImageProxy {
        private final String path;
        private RealImage realImage;

        public ImageProxy(String path) {
            this.path = path;
        }

        public String display() {
            if (realImage == null) {
                realImage = new RealImage(path);
            }
            return realImage.display();
        }
    }

    public static boolean demo() {
        RealImage.reset();
        ImageProxy proxy = new ImageProxy("chart.png");
        String first = proxy.display();
        String second = proxy.display();
        return first.equals(second) && RealImage.loadCount() == 1;
    }
}

File: csharp/patterns/src/PrototypePatterns/Structural/Proxy/ProxyPattern.cs
namespace PrototypePatterns.Structural.Proxy;

public sealed class RealImage
{
    private static int _loadCount;
    private readonly string _path;

    public RealImage(string path)
    {
        _path = path;
        _loadCount++;
    }

    public static int LoadCount => _loadCount;

    public static void Reset() => _loadCount = 0;

    public string Display() => $"display:{_path}";
}

public sealed class ImageProxy(string path)
{
    private RealImage? _realImage;

    public string Display()
    {
        _realImage ??= new RealImage(path);
        return _realImage.Display();
    }
}

public static class ProxyPattern
{
    public static bool Demo()
    {
        RealImage.Reset();
        var proxy = new ImageProxy("chart.png");
        var first = proxy.Display();
        var second = proxy.Display();
        return first == second && RealImage.LoadCount == 1;
    }
}
